design/open/hard-delete.md

```diff
@@ -0,0 +1,307 @@
+2. Add staging entry
+
+#### Delete Object
+
+1. If object staged
+   1. Read reference key
+   2. If not `committed` 
+      1. Change reference state to `deleted` and update `last_modified`
+
+#### Commit
+
+1. Mark all entries in staging area as `committed`
+2. Swap staging token
+3. Create commit
+4. Update branch commit ID
+
+#### Reset / Delete Branch
+
+When resetting a branch we throw all the uncommitted branch data, this operation happens asynchronously via `AsyncDrop`.
+We can leverage this asynchronous job to also check and perform the hard-delete
+1. Scan deleted staging token entries
+2. If entry is `tombstone` remove entry reference key
+3. Otherwise, modify reference state to `deleted` and update `last_modified`
+
+## 4. Staging Token States
+
+Track state on the staging token instead of objects - but keep tracking references (without state)
+References will be tracked in a list instead of an explicit entry per reference
+    
+    key    = <repo>/references/<physical_address>  
+    value  = list(staging tokens)
+
+For each staging token of a deleted / reset branches a deleted entry will be added
+
+    <repo>/staging/deleted/<staging_token>
+
+For each commit we will add the staging tokens to a committed entry.
+Adding a reference is done using `set if` and a retry mechanism to ensure consistency.
+
+    key    = <repo>/references/<physical_address>  
+    value  = list(staging tokens)
+
+### Flows
+
+#### Upload Object
+
+1. Write blob
+2. Read reference entry
+3. Perform `set if` on the reference with the additional staging token
+4. Add staged entry to database
+
+* Delete object will not remove the reference
+* References to objects are kept as long as staging token is not deleted or committed.
+
+#### Delete Object
+
+Deleted object will remain the same. We do not delete references in this flow.
+
+#### Stage Object
+Allow stage object operations only on addresses outside the repository namespace
+lakeFS will not manage the retention of these objects
+
+#### GetPhysicalAddress
+
+Provide a validation token with the issued physical address. The issued address can be used only once, thus
+we can assume the given address is not committed or referenced anywhere else.
+Once an issued address was used by LinkPhysicalAddress, we will delete the token ID from the list of valid tokens.
+
+1. Issue a JWT token with the physical address
+2. Save token ID in db
+3. return token with physical address to the user 
+
+#### LinkPhysicalAddress
+
+1. Given physical address, and token
+   1. Check token validity
+   2. If token valid
+      1. Remove token ID
+      2. Add reference
+      3. Add entry in staging area
+
+#### CopyObject
+
+The new flow ensures that we do not add a reference for a committed object after we removed all it's references as part of 
+the background delete job, and preventing the accidental deletion of committed objects.
+
+1. While retry not exhausted
+   1. Read reference list for address, if list is empty
```

Comment:
"List is empty" - does that include a key not found, or ONLY key found with empty list?

Reply:
Updated the background job and the copy flow to be more clear